### ISSUE TITLE
DistributedAdadeltaOptimizer Allreduce locks up whole network

### ISSUE THREAD

**Issue by vitalyli** [uber/horovod#343]
Trying to run it with NCCL2 on 32 P100 GPUs; it seem to be quite unstable and after a while of running, it hits this warning, after which the whole network locks up and nothing is moving after that.
Any way to debug it or avoid it? It has something to do with training data intake. At first I noticed that if
one of the processes is slower to pump the tensors through, then it has a tendency to lock up like that. 
The lockup starts with below warning, after which it never recovers.

**Comment by vitalyli** [comment]
I found that this lock up was caused by starting a test cycle run after N epochs.
In the log I see that starting test/eval loop immediately cause this lock up, from 
which it never recovered. The same approach worked ok with Tensorflow PS and Worker
setup, so I'm curious how to avoid this from crippling it?

Starting test tensor pump from rank: 0.. 
/data01/fstore/dev/test/part-00094
/data01/fstore/dev/test/part-00086
...
1530079544.343966 rank: 0, batch:0
WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. Stalled ops: OptimizerScope/DistributedAdadeltaOptimizer_Allreduce/HorovodAllreduce_OptimizerScope_gradients_FcnScope_MatMul_grad_tuple_control_dependency_1_0 [ready ranks: 1, 2, 3, 4, 7, 8, 9, 10, 11, 12, 18, 22, 23, 24, 26, 28, 29, 30, 31, 5, 6, 13, 14, 15, 16, 17, 19, 20, 21, 25, 27],
